components/hab/src/cli/hab/pkg.rs

```diff
@@ -437,3 +425,130 @@ pub enum Pkg {
+    ///
+    /// Run `hab pkg export <exporter> --help` to see exporter specific help. Example commands:
+    /// `hab pkg export tar --help`, `hab pkg export container core/redis --engine=buildah`
+    ///
+    /// Note: Not all exporters are available on all platforms.
+    #[structopt(parse(from_os_str))]
+    format_and_args: Vec<OsString>,
+}
+
+/// We must rely on manually parsing the export format and exporter arguments. The clap option
+/// `TrailingVarArg` allows capturing all arguments following a **positional** argument (ie the
+/// argument cannot be preceded by -- or -). Consider the following example commands:
+///
+/// 1. `hab pkg export container core/redis`
+/// 2. `hab pkg export container core/redis --multi-layer`
+/// 3. `hab pkg export container --multi-layer core/redis`
+///
+/// If we had a separate positional argument or subcommand for the export format examples #1 and #2
+/// would work. However, #3 would not work because `--multi-layer` does not look like a positional
+/// argument. To fix this we capture the export format as a positional argument.
+impl ExportCommand {
+    pub fn format(&self) -> Result<ExportFormat, ClapError> {
+        if let Some(format) = self.format_and_args.get(0) {
+            let format = format.to_string_lossy();
+            format.parse().map_err(|_| {
+                              ClapError::with_description(&format!("Failed to parse `{}` as an \
+                                                                    export format. [possible \
+                                                                    values: {}] ",
+                                                                   format,
+                                                                   ExportFormat::possible_values()),
+                                                          ClapErrorKind::InvalidValue)
+                          })
+        } else {
+            Err(ClapError::with_description(&format!("No export format was specified. [possible \
+                                                      values: {}] ",
+                                                     ExportFormat::possible_values()),
+                                            ClapErrorKind::MissingRequiredArgument))
+        }
+    }
+
+    pub fn args(&self) -> &[OsString] {
+        if self.format_and_args.is_empty() {
+            &[]
+        } else {
+            &self.format_and_args[1..]
+        }
+    }
+}
+
+/// This struct can only be constructed within this module because it has a private field. All
+/// `ExportFormat` variants contain an instance of this types. This effectively "seals" construction
+/// of `ExportFormat` to this module. Construction should only happen in `FromStr` which uses
+/// conditional compilation to only allow constructing platform supported `ExportFormat`s. By
+/// "sealing" construction of this enum we avoid the need to litter conditional compilation wherever
+/// `ExportFormat` is used.
```

Comment:
This is pretty slick; I'm a big fan of leveraging the type system like this to make the code cleaner and safer :+1: